src/miner.cpp

```diff
@@ -351,332 +385,387 @@ CBlockTemplate* CreateNewBlock(const CChainParams& chainparams, const MinerAddre
 
     // If we're given a coinbase tx, it's been precomputed, its fees are zero,
     // so we can't include any mempool transactions; this will be an empty block.
-    bool nonEmptyBlock = !next_cb_mtx;
+    blockFinished = blockFinished && next_cb_mtx;
 
-    // Largest block you're willing to create:
-    unsigned int nBlockMaxSize = GetArg("-blockmaxsize", DEFAULT_BLOCK_MAX_SIZE);
-    // Limit to between 1K and MAX_BLOCK_SIZE-1K for sanity:
-    nBlockMaxSize = std::max((unsigned int)1000, std::min((unsigned int)(MAX_BLOCK_SIZE-1000), nBlockMaxSize));
+    LOCK2(cs_main, mempool.cs);
+    CBlockIndex* pindexPrev = chainActive.Tip();
+    nHeight = pindexPrev->nHeight + 1;
+    uint32_t consensusBranchId = CurrentEpochBranchId(nHeight, chainparams.GetConsensus());
 
-    // How much of the block should be dedicated to high-priority transactions,
-    // included regardless of the fees they pay
-    unsigned int nBlockPrioritySize = GetArg("-blockprioritysize", DEFAULT_BLOCK_PRIORITY_SIZE);
-    nBlockPrioritySize = std::min(nBlockMaxSize, nBlockPrioritySize);
+    // -regtest only: allow overriding block.nVersion with
+    // -blockversion=N to test forking scenarios
+    if (chainparams.MineBlocksOnDemand())
+        pblock->nVersion = GetArg("-blockversion", pblock->nVersion);
 
-    // Minimum block size you want to create; block will be filled with free transactions
-    // until there are no more or the block reaches this size:
-    unsigned int nBlockMinSize = GetArg("-blockminsize", DEFAULT_BLOCK_MIN_SIZE);
-    nBlockMinSize = std::min(nBlockMaxSize, nBlockMinSize);
+    pblock->nTime = GetTime();
+    const int64_t nMedianTimePast = pindexPrev->GetMedianTimePast();
+    CCoinsViewCache view(pcoinsTip);
+
+    SaplingMerkleTree sapling_tree;
+    assert(view.GetSaplingAnchorAt(view.GetBestAnchor(SAPLING), sapling_tree));
```

Comment:
Not for this PR: it would be nice for `GetSaplingAnchorAt` to return std::optional instead.

Reply:
Also from @daira: why aren't we also getting the Orchard tree here?

Reply:
We need the Sapling tree here in order to correctly fill in the block header before Heartwood activation (when the header field currently named `hashBlockCommitments` was set to `hashFinalSaplingRoot`). We don't need Orchard here because from NU5 onward `hashBlockCommitments` commits to the tree roots as of the _start_ of the block via `hashChainHistoryRoot`, which can be fetched directly from the view.